### ISSUE TITLE
Float to int conversion loses sign

### ISSUE THREAD

**Issue by rotu** [nim-lang/Nim#20102]
Converting a large positive floating point number to an integer can result in a large negative integer.

### Example

```nim
let l = int.high
echo l
echo l.float.int
```
https://play.nim-lang.org/#ix=45QL

### Current Output
please check whether the problem still exists in git head before posting,
see [rebuilding the compiler](https://nim-lang.github.io/Nim/intern.html#rebuilding-the-compiler).
```
9223372036854775807
-9223372036854775808
```

### Expected Output
something like
```
9223372036854775807
9223372036854775807
```
or
```
9223372036854775807
9223372036854774784
```

### Additional Information
```
nim --version
Nim Compiler Version 1.6.6 [Windows: amd64]
Compiled at 2022-05-05
Copyright (c) 2006-2021 by Andreas Rumpf
```

**Comment by demotomohiro** [comment]
I got same result in C lang:
https://wandbox.org/permlink/QzZiIKrp1Tv1msxq

It seems converting int to float result in larger value because it is lossy convertion.
Converting float value larger than int.high result in negative value because it is undefined behavior?
Maybe Nim should throw `OverflowDefect` when converting float value larger than `int.high`.

**Comment by rotu** [comment]
Yeah, it's UB since the floating point is out of range of an int. I think an `OverflowDefect` would be appropriate. I'm also *very* surprised that trying to convert `Inf` and `NaN` throws an error at compile time but not at runtime.

**Comment by ringabout** [comment]
Rust outputs

```rust
fn main() {
    let x = 9223372036854775807_i64;
    let y = x as f64;
    let z = y as i64;
    println!("{} {} {}", x, y, z);
}
```

```
9223372036854775807
9223372036854776000
9223372036854775807
```
